New theme for documentation

alexisthual opened the pull request:
This PR resolves #3122 and changes the documentation theme to [furo](https://github.com/pradyunsg/furo/) for a better looking, more readable and accessible website!

You can preview these changes (without running the examples) by pulling this branch and running:
```
python -m pip install -r requirements-build-docs.txt
cd doc
make clean && make html-noplot
firefox _build/html/index.html
```

The main changes are:

- [x] add a "Quickstart" page which mirrors `./README.md`
- [x] documentation sidebar and navigation
  - [x] display examples' subsections in the sidebar [^1]
    - [x] open and merge PR allowing [nested sections in `sphinx-gallery`](https://github.com/sphinx-gallery/sphinx-gallery/issues/855)
  - [x] shorter names for doc sections and subsections (like "Contributing", "Introduction", "Maintenance", "API References", "Changelog", etc)
  - [x] create individual page for each module (new files are in `doc/modules`)
  - [x] redirect links from `doc/modules/references.rst` to new individual pages or just leave this file for now (I added a warning saying this part of the doc is no longer maintainted)
  - [x] move most basic tutorials to a new `examples/00_tutorials` folder for more consistent site structure (**this will break some existing urls**)
  - [x] redirect urls of previous tutorials to new tutorials
- [x] better landing page
- [x] better looking tabs for installation (resolves #3107)
- [x] better looking binder buttons
- [x] favicon with nilearn logo
- [x] make the logo compatible with light and dark themes
- [x] add new dependencies to `requirements-build-docs.txt` (`furo`, `recommonmark`, `sphinx-design`)
- [x] remove custom components (from `tune_toc.rst` and `big_toc_css.rst`)
- [x] enable markdown (we might want to get rid of this)
- [x] customise default thumbnail for examples gallery

Before deploying these changes, we should:
- [ ] re-read the whole documentation (I'll list the sections and assign names of people willing to participate!)
- [x] check that the examples build fine

[^1]: `furo` generates two sidebars: the left one shows the website's structure, the right one shows current document's structure. The one to the left is a sphinx table of content generated with `titles_only=True`, [which we can't change](https://github.com/pradyunsg/furo/discussions/354). This means that the left sidebar represents the exact site structure. If we want examples to be grouped in subsections in the sidebar, we need to create a new nested structure similar to what I did for `doc/modules`. However, examples are generated using `sphinx-gallery`, so we can either wait for [a `sphinx-gallery` issue mentioning nested sections](https://github.com/sphinx-gallery/sphinx-gallery/issues/855) to be solved or generate an individual gallery for each subsection (I'll try the first option quickly and jump to the second if it's not obvious).

alexisthual commented on doc/modules/connectome.rst:
```diff
@@ -0,0 +1,32 @@
+   :toctree: generated/
+   :template: class.rst
+
+   ConnectivityMeasure
```
I think the lines you highlight were manually written (these lines simply from from `doc/module/references.rst`, which I split in several files - one for each module). It's the definitions of modules' classes and functions which are automatically generated, no?

bthirion replied:
Sorry, I looked too superficially.

alexisthual commented:
We discussed having a live preview of these changes to proofread them before deploying.

I created [a repo](https://github.com/nilearn-doc-dev/nilearn-doc-dev.github.io) with the dev build so that we can see it using github pages.
It is accessible here: https://nilearn-doc-dev.github.io/index.html
However, it seems that css files from `_static` cannot be queried correctly, I don't know why :shrug:

GaelVaroquaux commented:
> However, it seems that css files from _static cannot be queried correctly, I don't know why 🤷

Maybe it's a case of github's jekyll engine interfeering (in jekyll, I
think that files starting with an underscore have a special meaning.

You can add an empty file named ".nojekyll" in the root of the html repo
to try to fix the problem:
https://github.blog/2009-12-29-bypassing-jekyll-on-github-pages/

tbng commented:
The new theme looks neat! It renders well on my mobile phone using firefox mobile as well.

I'm not sure about the content bar (column?) on the right side though. Maybe a better place is to also place it on the left -- although I guess this requires tuning a lot from the theme and not desireable. 

Also for now the Featured examples links on the main page does not work (all links to the Quick Start page).

alexisthual commented:
Sooo, I still need to make minor changes to the code, but I think we can start reading the documentation!

Current version is accessible here: https://nilearn-doc-dev.github.io

As discussed during the last nilearn dev meeting, I'll assign some parts of the documentation to read, but feel free to change these assignments! (and to tag people who might be interested)

Examples
- [ ] Basic tutorials @alexisthual 
- [ ] Visualization of brain images @alexisthual 
- [ ] Decoding @htwangtw  
- [ ] Functional connectivity @htwangtw 
- [ ] GLM first level @tsalo 
- [ ] GLM second level @tsalo 
- [ ] Manipulating brain volumes @emdupre 
- [ ] Advanced statistical analysis @emdupre 

Other
- [ ] Main page @jeromedockes 
- [ ] Quickstart @jeromedockes 
- [ ] User guide @jeromedockes (there might be a lot to say about this one)
- [ ] API references @NicolasGensollen from `connectivity` to `interfaces` included, @bthirion the rest
- [ ] All sections from the Development part in the sidebar (Migration, Contributing, etc) @tbng if you want :slightly_smiling_face:  

Our main goals are:
1. check that generated pages look ok
2. quickly check that links seem correct
3. check for parts with custom css / js etc
4. try to unify style. I encountered many different styles to add notes, warnings, etc. I suggest we stick to [admonition items from `furo`](https://pradyunsg.me/furo/kitchen-sink/admonitions/)

tsalo commented:
Everything in the first- and second-level GLM examples looks good to me. I saw the table of contents issue, but it looks like you're already aware of that. I do feel like the examples sections of the reference pages have a lot of wasted space by having the examples listed vertically.

alexisthual commented:
> Everything in the first- and second-level GLM examples looks good to me. I saw the table of contents issue, but it looks like you're already aware of that. I do feel like the examples sections of the reference pages have a lot of wasted space by having the examples listed vertically.

Sorry, currently, what issue do you see with the table of content? (it's supposed to be solved haha)
Also, what do you call the "reference pages"? Could you please send screenshots of the wasted space you mention? :slightly_smiling_face: 

Besides, I just found a broken link by chance in https://nilearn-doc-dev.github.io/auto_examples/04_glm_first_level/plot_fixed_effects.html#glm-estimation (link to the nistat's documentation), could you do a quick pass to check there aren't any of these? Thank you!

tsalo commented:
>Sorry, currently, what issue do you see with the table of content?

This is what I'm seeing:

![image](https://user-images.githubusercontent.com/8228902/157920644-b694e138-d696-464e-902a-c14839b1ade6.png)

It's still present in https://nilearn-doc-dev.github.io/auto_examples/04_glm_first_level/plot_bids_features.html#sphx-glr-auto-examples-04-glm-first-level-plot-bids-features-py and https://nilearn-doc-dev.github.io/auto_examples/04_glm_first_level/plot_first_level_details.html#sphx-glr-auto-examples-04-glm-first-level-plot-first-level-details-py, at least.

>Also, what do you call the "reference pages"? Could you please send screenshots of the wasted space you mention?

I'm referring to the API documentation. For example, on [the page for FirstLevelModel](https://nilearn-doc-dev.github.io/modules/generated/nilearn.glm.first_level.FirstLevelModel.html#nilearn.glm.first_level.FirstLevelModel), the examples look like this:

![image](https://user-images.githubusercontent.com/8228902/157921144-6f47efad-2ff1-4715-bd34-3d63f307714d.png)

In the official documentation, they look like this:

![image](https://user-images.githubusercontent.com/8228902/157921171-24f8736a-adeb-47b5-bba8-2a224e62e070.png)

It's not a big deal, but there ends up being a lot of empty space with that organization, and some functions or classes that show up in a lot of examples are going to have _very_ long example sections.

>Besides, I just found a broken link by chance in https://nilearn-doc-dev.github.io/auto_examples/04_glm_first_level/plot_fixed_effects.html#glm-estimation (link to the nistat's documentation), could you do a quick pass to check there aren't any of these?

I did notice that, but it's also present in the official documentation, so it's a typo in the actual RST page rather than a problem with the new theme.

alexisthual commented:
Thanks a lot, that's exactly the kind of feedback I need :)

htwangtw commented:
I have encountered some similar issues so here's a concise summary:

- This page has the same table of content error mentioned above, also a lack of pretty thumb nail:
`examples/02_decoding/plot_haxby_glm_decoding.py`

- Minor cosmetic issue - explanation text rendered as comments:
`examples/03_connectivity/plot_seed_to_voxel_correlation.py`

- Minor cosmetic issue on the references:
`examples/03_connectivity/plot_sphere_based_connectome.py`
`examples/03_connectivity/plot_data_driven_parcellations.py`

alexisthual commented:
> * Minor cosmetic issue on the references:
>   `examples/03_connectivity/plot_sphere_based_connectome.py`
>   `examples/03_connectivity/plot_data_driven_parcellations.py`

What do you think would be a good way to display references? Can you give me a visual example please?

htwangtw commented:
> What do you think would be a good way to display references? Can you give me a visual example please?

Of course! After looking at some other pages that I am not in charge of, I think it's really me nitpicking. Feel free to ignore them!
It's up to you if you want to do it :smile: 

I was expecting the references to work like this page here, with proper internal link: [examples/03_connectivity/plot_signal_extraction](https://nilearn-doc-dev.github.io/auto_examples/03_connectivity/plot_signal_extraction.html#references)

![image](https://user-images.githubusercontent.com/13743617/157954450-ff581593-a8ba-4ac2-932f-7d5b9d41e516.png)

This one [03_connectivity/plot_sphere_based_connectome](https://nilearn-doc-dev.github.io/auto_examples/03_connectivity/plot_sphere_based_connectome.html) uses plain text rather than linked 

![image](https://user-images.githubusercontent.com/13743617/157953033-c0c9ba3e-b613-4dc8-8c63-dbc724ae570c.png)

This one [examples/03_connectivity/plot_data_driven_parcellations](https://nilearn-doc-dev.github.io/auto_examples/03_connectivity/plot_data_driven_parcellations.html) just looks a bit too stand out comparing to the original:

Original - 

![image](https://user-images.githubusercontent.com/13743617/157955717-07363d36-7102-4e77-8e19-ce9cecf22579.png)

New - 

![image](https://user-images.githubusercontent.com/13743617/157955528-5a792995-4780-4594-a1f3-de27c34b841e.png)

alexisthual commented:
Nice, I'll stick to the way [03_connectivity/plot_signal_extraction.py](https://nilearn-doc-dev.github.io/auto_examples/03_connectivity/plot_signal_extraction.html#references) handles references then!

NicolasGensollen commented:
@alexisthual I had a very quick look at the API pages and I think they look great! :+1:  

There are still some differences in formatting between modules but this isn't related to this PR but rather to the unfinished homogenization of the docstrings accross Nilearn.

One small rendering issue I could spot is that sometimes (I think when there is a column on the right), long names get truncated. See for example:

![new_theme_2](https://user-images.githubusercontent.com/2639645/158054441-4a078a29-afe5-48ac-94bc-21a321156533.png)

or

![new_theme_1](https://user-images.githubusercontent.com/2639645/158054445-c80535f0-7f96-42bf-9bf1-c65933e712f3.png)

bthirion commented:
Weirdly, the documentation has a `nilearn.input_data`, yet not `nielarn.maskers`: it looks it does not have yet the latest evolution of the documentation.

emdupre commented on examples/05_glm_second_level/README.txt:
```diff
@@ -1,5 +1,6 @@
-GLM : Second level analysis examples
--------------------------------------
+===========================
+GLM : Second level analysis
```
I think we need to remove the space before the colon, here, to standardize with `GLM: First Level Analysis` !

```suggestion
GLM: Second level analysis
```

emdupre commented on doc/manipulating_images/masker_objects.rst:
```diff
@@ -8,10 +8,6 @@ This chapter introduces the maskers: objects that go from
 neuroimaging volumes, on the disk or in memory, to data matrices, eg of
 time series.
 
-.. contents:: **Chapters contents**
```
Trying to summarize all of my changes related to just this file in one place:

- We have a mix of admonition types in this file. Several of the `topic` admonitions are a running reference to sklearn, so I think they're OK to keep, but [this one](https://github.com/nilearn/nilearn/pull/3125/files#diff-058459b3f210e511ba10547afb134091ee23ba9205a7dbd19224c95ae8cbec85R205) should probably be a warning since it's advertising an upcoming change (which iirc will actually be a deprecation of int64).
- This [linking is broken](https://github.com/nilearn/nilearn/pull/3125/files#diff-058459b3f210e511ba10547afb134091ee23ba9205a7dbd19224c95ae8cbec85R202) to the `NiftiMasker` class name
- This [scale value](https://github.com/nilearn/nilearn/pull/3125/files#diff-058459b3f210e511ba10547afb134091ee23ba9205a7dbd19224c95ae8cbec85R142) is now making the image look too small with the current CSS. Could we drop or increase it ?

alexisthual replied:
Sorry your links don't seem to take me to modified lines (they all lead me to the page of changed files, but not to any file in particular). I know this was a long time ago already, sorry sorry :scream_cat:

emdupre replied:
Not sure if it will link anymore since you pulled in new changes :disappointed:  I won't re-generate all the links unless you have trouble finding the line in this file -- hopefully it's clear at least from the text what I'm referring to !

emdupre commented:
> User guide @jeromedockes (there might be a lot to say about this one)

Just to be clear : I'm going through the user guide for "Manipulation brain volumes with nilearn" since it so cleanly maps to the examples I'm already assigned :smile_cat: I agree that this is a big portion of the docs, so I'm happy to take another part too -- I'll preemptively volunteer for "Advanced usage: manual pipelines and scaling up." But let me know if there are any concerns, here !

alexisthual commented:
> > User guide @jeromedockes (there might be a lot to say about this one)
> 
> Just to be clear : I'm going through the user guide for "Manipulation brain volumes with nilearn" since it so cleanly maps to the examples I'm already assigned smile_cat I agree that this is a big portion of the docs, so I'm happy to take another part too -- I'll preemptively volunteer for "Advanced usage: manual pipelines and scaling up." But let me know if there are any concerns, here !

Great, let's assign all parts of the User Guide then :slightly_smiling_face: 
- [ ] Introduction @jeromedockes 
- [ ] Decoding and MVPA @jeromedockes 
- [ ] Functional connectivity @jeromedockes 
- [ ] Plotting brain images @alexisthual 
- [ ] Analyzing fMRI using GLMs @tsalo 
- [ ] Manipulating brain volumes @emdupre 
- [ ] Advanced usage @emdupre

> Weirdly, the documentation has a `nilearn.input_data`, yet not `nielarn.maskers`: it looks it does not have yet the latest evolution of the documentation.

I just pulled upstream main and will push a new dev version of the documentation soon!

jeromedockes commented:
the main page and quickstart look great.
I'm not sure about the choice of featured examples, I think some could be removed, but that's probably a discussion for another issue

in the quick start,
- some dependencies seem to be missing from the list: lxml, requests. (that is not really related to this pr, we should list dependencies in setup.cfg only and link to it in the readme)
- link to how to contribute page is an absolute URL, most other links I saw in main page & user guide use relative urls

also in the [examples page section 3](https://nilearn-doc-dev.github.io/auto_examples/index.html#functional-connectivity) a sphinx command is shown verbatim ":term:`Dictionary learning`"
and in https://nilearn-doc-dev.github.io/plotting/index.html images appear squished (too narrow)

tsalo commented:
I've gone through the GLM user guide section. It looks great, and there are no real issues with the rendering. Here are a few things that I thought might be worth changing:

1. I vastly prefer the new admonition style, but would it be possible to keep the font size consistent? The [note here](https://nilearn-doc-dev.github.io/glm/glm_intro.html#what-is-fmri) has a smaller font than the main text. I think the "Page Summary" admonition could have a bolder color as well.
2. Would it be possible to implement tooltips for glossary items? It's not a big deal if not, but they could be nice to have.
3. Some images are left-aligned, while others are centered. For example, the image [here](https://nilearn-doc-dev.github.io/glm/first_level_model.html#computing-contrasts) is on the left, while the one in [the next section](https://nilearn-doc-dev.github.io/glm/first_level_model.html#extracting-predicted-time-series-and-residuals) is centered.

jeromedockes commented on doc/introduction.rst:
```diff
@@ -89,8 +85,72 @@ Why is machine learning relevant to NeuroImaging? A few examples!
+
+  .. tab-item:: Pip install
+
+      **1. Download and install 64-bit Anaconda**
```
also python and nilearn work fine on 32 bit architectures

alexisthual replied:
It's good that users use a custom python env though, don't you think? Do you want to recommend miniconda instead?

alexisthual replied:
> the main page and quickstart look great. I'm not sure about the choice of featured examples, I think some could be removed, but that's probably a discussion for another issue

Yes, let's keep it for another PR maybe :)

> * some dependencies seem to be missing from the list: lxml, requests. (that is not really related to this pr, we should list dependencies in setup.cfg only and link to it in the readme)

It has be resolved since you posted this I believe

> * link to how to contribute page is an absolute URL, most other links I saw in main page & user guide use relative urls

This is because `quickstart` is merely a copy of the `readme.md`, which is our github frontpage, so urls need to be absolute.

> also in the [examples page section 3](https://nilearn-doc-dev.github.io/auto_examples/index.html#functional-connectivity) a sphinx command is shown verbatim ":term:`Dictionary learning`" and in https://nilearn-doc-dev.github.io/plotting/index.html images appear squished (too narrow)

I removed the term tag in the title. Besides, I think I fixed the squished images problem.

jeromedockes commented on doc/introduction.rst:
```diff
@@ -89,8 +85,72 @@ Why is machine learning relevant to NeuroImaging? A few examples!
+      64-bit scientific Python distribution like
+      `Anaconda <https://www.anaconda.com/download/>`_ (or `PythonXY <http://python-xy.github.io/>`_).
+
+      Nilearn requires a Python installation and the following
```
here we should point to the list of dependencies we already have elsewhere instead of having a copy of the list

jeromedockes commented on doc/introduction.rst:
```diff
@@ -89,8 +85,72 @@ Why is machine learning relevant to NeuroImaging? A few examples!
+      dependencies: ipython, scipy, scikit-learn, joblib, matplotlib,
+      nibabel and pandas.
+
+      Once conda is installed, create and activte
```
I think the standard library `venv` is a better option than conda

jeromedockes commented on doc/introduction.rst:
```diff
@@ -89,8 +85,72 @@ Why is machine learning relevant to NeuroImaging? A few examples!
+
+      .. code-block::
+
+         python -m pip install -U --user nilearn
```
why `--user` if we are in a virtual env?

alexisthual replied:
I guess this was a relic from the past. I got rid of it!

jeromedockes reviewed (commented):
minor comments on the user guide / introduction; overall it displays and reads well

alexisthual commented:
> 1. I vastly prefer the new admonition style, but would it be possible to keep the font size consistent? The [note here](https://nilearn-doc-dev.github.io/glm/glm_intro.html#what-is-fmri) has a smaller font than the main text. I think the "Page Summary" admonition could have a bolder color as well.

I agree with you that admonition font size shouldn't be noticeably smaller. I increased it to `100%`.

> 2. Would it be possible to implement tooltips for glossary items? It's not a big deal if not, but they could be nice to have.

I don't think it'd be very easy, but that's a cool idea indeed! Maybe for another PR?

> 3. Some images are left-aligned, while others are centered. For example, the image [here](https://nilearn-doc-dev.github.io/glm/first_level_model.html#computing-contrasts) is on the left, while the one in [the next section](https://nilearn-doc-dev.github.io/glm/first_level_model.html#extracting-predicted-time-series-and-residuals) is centered.

Indeed, rst `figure`s are centered by default, while rst `image`s are left-aligned and could be centered manually (with `:align: center` for instance). I propose to leave these defaults as they are for now, as we don't yet have a clear policy on how to handle this :)

alexisthual commented on doc/introduction.rst:
```diff
@@ -89,8 +85,72 @@ Why is machine learning relevant to NeuroImaging? A few examples!
-   :file: install_doc_component.html
+.. tab-set::
+
+  .. tab-item:: Pip install
```
Done!

alexisthual commented:
Soooo, a new version of `sphinx-gallery` with the changes we need just came out :tada: https://github.com/sphinx-gallery/sphinx-gallery/pull/981
The CI now runs without error, so I would advocate to merge this soon. Indeed, some parts of the documentation might benefit from a more careful read, but I think we proofread a great part of the website already and feel confident it won't crash. Moreover, I am deploying an updated preview of the docs here for ultimate tests: https://nilearn-doc-dev.github.io/

However, while we're here, we might as well plan some changes for the documentation.
1. In my humble opinion, we could do a much better job with the User Guide. In particular:
2. It is filled with interesting examples, references and pieces of code, but sometimes feels like a patchwork. I think it would greatly benefit from being harmonized by someone who could spend the time to read it entirely. Maybe we could make that a nilearn dev days project for new-comers?
3. Moreover, I'd be in favor of giving it a more ambitious name: To me, it almost feels like a short textbook on fMRI, why not name it this way?
4. Its main page looks a bit rough haha, but we can change that later.
5. Oftentimes, examples are ordered in a suboptimal way within sections. I think this is an important aspect of how we introduce core features of the library to users, and ultimately we should find a way to polish a more straight-to-the-point interface.

bthirion commented:
Thx !  Since your suggestions are big steps, shouldn't we proceed by first merging that one and then reviewing more in detail the documentation ?

ymzayek commented:
I agree with those points @alexisthual. I am also in favor of merging this first and then working through them. I trust it has been reviewed sufficiently but I can have a look today at the updated docs just to check it overall.

ymzayek commented on setup.cfg:
```diff
@@ -30,10 +30,15 @@ doc =
 	coverage
 	flake8
+	furo
+	kaleido
```
```suggestion
```
This is redundant because kaleido is included with `%(plotly)s` in line 30

ymzayek reviewed (commented):
Hi @alexisthual, there are a few small things I think we can try to fix before merging. 

- The note [on this page](https://nilearn-doc-dev.github.io/auto_examples/06_manipulating_images/plot_nifti_labels_simple.html) is in italics. Reformat to normal like others.
- There is an error [on this page](https://nilearn-doc-dev.github.io/auto_examples/07_advanced/plot_advanced_decoding_scikit.html)
- Plots are rendered very small [here](https://nilearn-doc-dev.github.io/auto_examples/06_manipulating_images/plot_extract_rois_statistical_maps.html) and also some plots [here](https://nilearn-doc-dev.github.io/auto_examples/04_glm_first_level/plot_first_level_details.html#sphx-glr-auto-examples-04-glm-first-level-plot-first-level-details-py)
- Isn't this [module](https://nilearn-doc-dev.github.io/modules/masking.html) supposed to be called `maskers` instead of `masking`

Also ok if you think some things are better fixed separately in a different PR after merging this. Let me know.

ymzayek commented:
>Isn't this [module](https://nilearn-doc-dev.github.io/modules/masking.html) supposed to be called `maskers` instead of `masking`

Actually this is reflecting the current documentation and can stay as is. 

But this related comment should be addressed: https://github.com/nilearn/nilearn/pull/3125#issuecomment-1066657685

alexisthual commented:
Thanks for your comments @ymzayek, I addressed all of them in my last commit.

Point by point:
- I had to modify `./nilearn/reporting/data/html/report_body_template_niftilabelsmasker.html` ; more generally, this folder is filled with HTML / CSS content I don't really like haha
- removing the table of content removed the error
- fixed this with a minor addition to our custom CSS file to handle image lists generated by sphinx-gallery
- good catch, a section was actually missing in the API references (I had split `reference.rst` for more readability, and missed that `input_data` was renamed to `maskers`)

jeromedockes commented:
> 5\. Oftentimes, examples are ordered in a suboptimal way within sections. I think this is an important aspect of how we introduce core features of the library to users, and ultimately we should find a way to polish a more straight-to-the-point interface.

see also #3195

ymzayek reviewed (approved):
Thanks @alexisthual for the quick changes! For me I think it is in a good state to merge.

bthirion commented on examples/03_connectivity/plot_sphere_based_connectome.py:
```diff
@@ -16,6 +16,7 @@
 
 [2] Dosenbach N.U., Nardos B., et al. "Prediction of individual brain maturity
 using fMRI.", 2010, Science 329, 1358-1361.
+>>>>>>> 31737ae62fafe9c9d4f12ac0ff5f92a43dcb93aa
```
line to be removed.

bthirion reviewed (commented):
I haven't identified any big issue in the rendering. Thx a lot !

bthirion reviewed (approved):
Many thx for modernizing the doc !

ymzayek commented:
When everything passes I will have a quick look at the artifacts and then merge.

ymzayek reviewed (commented):
@alexisthual I really think it's almost there! The fix in `AUTHORS.rst` should do the trick. See below.

ymzayek commented:
Thanks @alexisthual and all reviewers for your work on this. Merging!
